### PR TITLE
Adding 64-byte access support for CXL.mem

### PR DESCRIPTION
**Pull request by zhxq**
Adding 64-byte access support for CXL.mem

### CODE REVIEW

File: hw/cxl/cxl-host.c [hunk 1]
```diff
@@ -210,36 +209,36 @@ static MemTxResult cxl_read_cfmws(void *opaque, hwaddr addr, uint64_t *data,
...
     d = cxl_cfmws_find_device(fw, addr);
     if (d == NULL) {
+        trace_cxl_debug_message("CXL device not found");
         *data = 0;
         /* Reads to invalid address return poison */
         return result;
     }
-    type = object_get_typename(OBJECT(d));
 
-    g_assert(addr < 0x8000000);
```
Comment (kyeyoon-eeum) [inline hw/cxl/cxl-host.c:219]: Why remove assert?

File: hw/cxl/cxl-host.c [hunk 2]
```diff
@@ -248,26 +247,27 @@ static MemTxResult cxl_write_cfmws(void *opaque, hwaddr addr,
 
     d = cxl_cfmws_find_device(fw, addr);
     if (d == NULL) {
+        trace_cxl_debug_message("CXL device not found");
         /* Writes to invalid address are silent */
         return result;
     }
-    type = object_get_typename(OBJECT(d));
-
-    g_assert(addr < 0x8000000);
```
Comment (kyeyoon-eeum) [inline hw/cxl/cxl-host.c:256]: Same here. Why remove assert?

File: hw/cxl/cxl-host.c [hunk 3]
```diff
@@ -355,7 +356,8 @@ void cxl_hook_up_pxb_registers(PCIBus *bus, CXLState *state, Error **errp)
 {
     /* Walk the pci busses looking for pxb busses to hook up */
     if (bus) {
-        QLIST_FOREACH(bus, &bus->child, sibling) {
+        QLIST_FOREACH(bus, &bus->child, sibling)
+        {
```
Comment (kyeyoon-eeum) [inline hw/cxl/cxl-host.c:360]: Did clang-format do this on its own? Move curly to the above line.

File: hw/pci-bridge/cxl_root_port.c [hunk 4]
```diff
@@ -64,6 +64,11 @@ typedef struct CXLRootPort {
 #define TYPE_CXL_ROOT_PORT "cxl-rp"
 DECLARE_INSTANCE_CHECKER(CXLRootPort, CXL_ROOT_PORT, TYPE_CXL_ROOT_PORT)
 
+static bool cxl_rw_buffer_inited[CXL_RW_NUM_BUFFERS] = { 0 };
+static hwaddr cxl_rw_buffer_page[CXL_RW_NUM_BUFFERS] = { 0 };
+static uint64_t cxl_rw_last_access_time[CXL_RW_NUM_BUFFERS] = { 0 };
+static uint8_t cxl_rw_buffer[CXL_RW_NUM_BUFFERS][CXL_MEM_ACCESS_UNIT];
```
Comment (kyeyoon-eeum) [inline hw/pci-bridge/cxl_root_port.c:70]: Let's use `struct`.

Indicate that it is for CXL.mem, e.g. `cxl_mem_rw_<...>`

File: hw/pci-bridge/cxl_root_port.c [hunk 5]
```diff
@@ -110,30 +125,79 @@ MemTxResult cxl_remote_cxl_mem_read(PCIDevice *d, hwaddr host_addr,
...
-    *data = *(uint64_t *)(cxl_packet->data);
+    *data = *(uint8_t *)(cxl_packet->data);
     release_packet_entry(tag);
 
     return MEMTX_OK;
 }
 
+uint64_t cxl_get_dest_cache(PCIDevice *d, hwaddr host_addr, MemTxAttrs attrs)
+{
+    // TODO: maybe a lock mechanism here?
```
Comment (kyeyoon-eeum) [inline hw/pci-bridge/cxl_root_port.c:140]: For comments in C, please use `/* */`.

File: hw/pci-bridge/cxl_root_port.c [hunk 6]
```diff
@@ -110,30 +125,79 @@ MemTxResult cxl_remote_cxl_mem_read(PCIDevice *d, hwaddr host_addr,
...
+    uint32_t cache_iter;
+    bool cache_hit = false;
+    uint64_t least_accessed_time = 0ULL - 1;
+    uint64_t least_accessed_candidate = 0;
+
+    for (cache_iter = 0; cache_iter < CXL_RW_NUM_BUFFERS; cache_iter++) {
```
Comment (kyeyoon-eeum) [inline hw/pci-bridge/cxl_root_port.c:146]: Let's call it `cache_idx` or `idx`.

`cache_iter` is more than a simple iterator looking at the below code.

File: hw/pci-bridge/cxl_root_port.c [hunk 7]
```diff
@@ -110,30 +125,79 @@ MemTxResult cxl_remote_cxl_mem_read(PCIDevice *d, hwaddr host_addr,
...
+            break;
+        }
+        if (cxl_rw_last_access_time[cache_iter] < least_accessed_time) {
+            least_accessed_time = cxl_rw_last_access_time[cache_iter];
+            least_accessed_candidate = cache_iter;
+        }
+    }
+    if (!cache_hit) {
+        // else evict one and use the freed one as cache
+        // Flush the existing to the backend
```
Comment (kyeyoon-eeum) [inline hw/pci-bridge/cxl_root_port.c:164]: /* */

File: hw/pci-bridge/cxl_root_port.c [hunk 8]
```diff
@@ -110,30 +125,79 @@ MemTxResult cxl_remote_cxl_mem_read(PCIDevice *d, hwaddr host_addr,
...
+        cache_iter = least_accessed_candidate;
+        cxl_remote_cxl_mem_write(d, cxl_rw_buffer_page[cache_iter] << 6,
+                                 &cxl_rw_buffer[cache_iter][0],
+                                 CXL_MEM_ACCESS_UNIT, attrs);
+        cxl_rw_buffer_page[cache_iter] = host_addr >> 6;
+        cxl_rw_last_access_time[cache_iter] =
+            qemu_clock_get_ns(QEMU_CLOCK_VIRTUAL);
+        // need to first read from the location without going through the cache
```
Comment (kyeyoon-eeum) [inline hw/pci-bridge/cxl_root_port.c:172]: /* */

File: hw/pci-bridge/cxl_root_port.c [hunk 9]
```diff
@@ -110,30 +125,79 @@ MemTxResult cxl_remote_cxl_mem_read(PCIDevice *d, hwaddr host_addr,
...
 
     return MEMTX_OK;
 }
 
+uint64_t cxl_get_dest_cache(PCIDevice *d, hwaddr host_addr, MemTxAttrs attrs)
+{
+    // TODO: maybe a lock mechanism here?
+    uint32_t cache_iter;
+    bool cache_hit = false;
+    uint64_t least_accessed_time = 0ULL - 1;
```
Comment (kyeyoon-eeum) [inline hw/pci-bridge/cxl_root_port.c:143]: Don't intentionally <over/under>flow an unsigned int. Use a MAX constant from <limits.h>

File: hw/cxl/cxl-host.c [hunk 10]
```diff
@@ -210,36 +209,36 @@ static MemTxResult cxl_read_cfmws(void *opaque, hwaddr addr, uint64_t *data,
...
-    g_assert(addr < 0x8000000);
-
-    if (g_strcmp0(type, "cxl-type1") == 0)
-        result = cxl_type1_read(d, addr + fw->base, data, size, attrs);
-    else if (g_strcmp0(type, "cxl-type2") == 0)
-        result =
-            cxl_host_type2_hcoh_read(d, addr + fw->base, data, size, attrs);
-    else if (g_strcmp0(type, "cxl-type3") == 0) {
-        if (cxl_is_remote_root_port(d)) {
+    if (cxl_is_remote_root_port(d)) {
```
Comment (kyeyoon-eeum) [inline hw/cxl/cxl-host.c:218]: Could you explain why this check needs to be added? How did it work without it before?

**Review by kyeyoon-eeum — CHANGES_REQUESTED** [review]

File: hw/cxl/cxl-host.c (hunk 3 above)
Comment (zhxq) [reply hw/cxl/cxl-host.c:360]: It was automatically done by clang-format.

File: hw/cxl/cxl-host.c (hunk 1 above)
Comment (zhxq) [reply hw/cxl/cxl-host.c:219]: Seems like casued by version conflict.

File: hw/cxl/cxl-host.c (hunk 2 above)
Comment (zhxq) [reply hw/cxl/cxl-host.c:256]: Seems like casued by version conflict.

**Pull request closed without merging**